File: training/load_csv.jl
using DataFrames, CSV


const CCLE = CSV.read("../gene_expression/tpm_values.csv", DataFrame)


function get_tpm(gene::String, cell_line::String)::Float64
    return CCLE[CCLE[!,Symbol("Column1")].==gene,Symbol(cell_line)][1]
end


function init_E1(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_E1] * get_tpm("EGFR", cell_line)
end


function init_E2(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_E2] * get_tpm("ERBB2", cell_line)
end


function init_E3(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_E3] * get_tpm("ERBB3", cell_line)
end


function init_E4(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_E4] * get_tpm("ERBB4", cell_line)
end


function init_G(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_G] * get_tpm("GRB2", cell_line)
end


function init_S(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_SHC1] * get_tpm("SHC1", cell_line) +
            p[C.a_SHC2] * get_tpm("SHC2", cell_line) +
            p[C.a_SHC3] * get_tpm("SHC3", cell_line) +
            p[C.a_SHC4] * get_tpm("SHC4", cell_line))
end


function init_I(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_PIK3CA] * get_tpm("PIK3CA", cell_line) +
            p[C.a_PIK3CB] * get_tpm("PIK3CB", cell_line) +
            p[C.a_PIK3CD] * get_tpm("PIK3CD", cell_line) +
            p[C.a_PIK3CG] * get_tpm("PIK3CG", cell_line))
end

function init_PTEN(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_PTEN] * get_tpm("PTEN", cell_line)
end

function init_R(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_RASA1] * get_tpm("RASA1", cell_line) +
            p[C.a_RASA2] * get_tpm("RASA2", cell_line) +
            p[C.a_RASA3] * get_tpm("RASA3", cell_line))
end


function init_O(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_SOS1] * get_tpm("SOS1", cell_line) +
            p[C.a_SOS2] * get_tpm("SOS2", cell_line))
end


function init_A(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_A] * get_tpm("GAB1", cell_line)
end


function init_Akt(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_AKT1] * get_tpm("AKT1", cell_line) +
            p[C.a_AKT2] * get_tpm("AKT2", cell_line))
end


function init_RsD(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_HRAS] * get_tpm("HRAS", cell_line) +
            p[C.a_KRAS] * get_tpm("KRAS", cell_line) +
            p[C.a_NRAS] * get_tpm("NRAS", cell_line))
end


function init_Raf(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_ARAF] * get_tpm("ARAF", cell_line) +
            p[C.a_BRAF] * get_tpm("BRAF", cell_line) +
            p[C.a_RAF1] * get_tpm("RAF1", cell_line))
end


function init_MEK(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_MAP2K1] * get_tpm("MAP2K1", cell_line) +
            p[C.a_MAP2K2] * get_tpm("MAP2K2", cell_line))
end


function init_T(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_T] * get_tpm("PTPN1", cell_line)
end


function init_CREBn(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_CREBn] * get_tpm("CREB1", cell_line)
end


function init_ERKc(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_MAPK1] * get_tpm("MAPK1", cell_line) +
            p[C.a_MAPK3] * get_tpm("MAPK3", cell_line))
end


function init_Elk1n(p::Vector{Float64}, cell_line::String)::Float64
    return p[C.scale_Elk1n] * get_tpm("ELK1", cell_line)
end


function init_RSKc(p::Vector{Float64}, cell_line::String)::Float64
    return (p[C.a_RPS6KA1] * get_tpm("RPS6KA1", cell_line) +
            p[C.a_RPS6KA2] * get_tpm("RPS6KA2", cell_line) +
            p[C.a_RPS6KA3] * get_tpm("RPS6KA3", cell_line) +
            p[C.a_RPS6KA6] * get_tpm("RPS6KA6", cell_line))
end


function tpm2ival!(p::Vector{Float64}, u0::Vector{Float64}, cell_line::String)
    p[C.PTEN] *= init_PTEN(p, cell_line)

    u0[V.E1] *= init_E1(p, cell_line)
    u0[V.E2] *= init_E2(p, cell_line)
    u0[V.E3] *= init_E3(p, cell_line)
    u0[V.E4] *= init_E4(p, cell_line)
    u0[V.G] *= init_G(p, cell_line)
    u0[V.S] *= init_S(p, cell_line)
    u0[V.I] *= init_I(p, cell_line)
    u0[V.R] *= init_R(p, cell_line)
    u0[V.O] *= init_O(p, cell_line)
    u0[V.A] *= init_A(p, cell_line)
    u0[V.Akt] *= init_Akt(p, cell_line)
    u0[V.RsD] *= init_RsD(p, cell_line)
    u0[V.Raf] *= init_Raf(p, cell_line)
    u0[V.MEK] *= init_MEK(p, cell_line)
    u0[V.T] *= init_T(p, cell_line)
    u0[V.CREBn] *= init_CREBn(p, cell_line)
    u0[V.ERKc] *= init_ERKc(p, cell_line)
    u0[V.Elk1n] *= init_Elk1n(p, cell_line)
    u0[V.RSKc] *= init_RSKc(p, cell_line)

    return p, u0
end


function mul_ratio2mcf7!(p::Vector{Float64}, u0::Vector{Float64}, cell_line_2::String)
    p[C.PTEN] *= init_PTEN(p, cell_line_2) / init_PTEN(p, "MCF7")

    u0[V.E1] *= init_E1(p, cell_line_2) / init_E1(p, "MCF7")
    u0[V.E2] *= init_E2(p, cell_line_2) / init_E2(p, "MCF7")
    u0[V.E3] *= init_E3(p, cell_line_2) / init_E3(p, "MCF7")
    u0[V.E4] *= init_E4(p, cell_line_2) / init_E4(p, "MCF7")
    u0[V.G] *= init_G(p, cell_line_2) / init_G(p, "MCF7")
    u0[V.S] *= init_S(p, cell_line_2) / init_S(p, "MCF7")
    u0[V.I] *= init_I(p, cell_line_2) / init_I(p, "MCF7")
    u0[V.R] *= init_R(p, cell_line_2) / init_R(p, "MCF7")
    u0[V.O] *= init_O(p, cell_line_2) / init_O(p, "MCF7")
    u0[V.A] *= init_A(p, cell_line_2) / init_A(p, "MCF7")
    u0[V.Akt] *= init_Akt(p, cell_line_2) / init_Akt(p, "MCF7")
    u0[V.RsD] *= init_RsD(p, cell_line_2) / init_RsD(p, "MCF7")
    u0[V.Raf] *= init_Raf(p, cell_line_2) / init_Raf(p, "MCF7")
    u0[V.MEK] *= init_MEK(p, cell_line_2) / init_MEK(p, "MCF7")
    u0[V.T] *= init_T(p, cell_line_2) / init_T(p, "MCF7")
    u0[V.CREBn] *= init_CREBn(p, cell_line_2) / init_CREBn(p, "MCF7")
    u0[V.ERKc] *= init_ERKc(p, cell_line_2) / init_ERKc(p, "MCF7")
    u0[V.Elk1n] *= init_Elk1n(p, cell_line_2) / init_Elk1n(p, "MCF7")
    u0[V.RSKc] *= init_RSKc(p, cell_line_2) / init_RSKc(p, "MCF7")

    return p, u0
end

File: python/load_csv.py
import pandas as pd
from name2idx import C, V


CCLE = pd.read_csv('../gene_expression/tpm_values.csv', index_col=0)


def get_tpm(gene, cell_line, source):
    return CCLE.loc[gene, cell_line]


def init_E1(x, cell_line, source):
    return x[C.w_E1] * get_tpm("EGFR", cell_line, source)


def init_E2(x, cell_line, source):
    return x[C.w_E2] * get_tpm("ERBB2", cell_line, source)


def init_E3(x, cell_line, source):
    return x[C.w_E3] * get_tpm("ERBB3", cell_line, source)


def init_E4(x, cell_line, source):
    return x[C.w_E4] * get_tpm("ERBB4", cell_line, source)


def init_G(x, cell_line, source):
    return x[C.w_G] * get_tpm("GRB2", cell_line, source)


def init_S(x, cell_line, source):
    return (x[C.w_SHC1] * get_tpm("SHC1", cell_line, source) +
            x[C.w_SHC2] * get_tpm("SHC2", cell_line, source) +
            x[C.w_SHC3] * get_tpm("SHC3", cell_line, source) +
            x[C.w_SHC4] * get_tpm("SHC4", cell_line, source))


def init_I(x, cell_line, source):
    return (x[C.w_PIK3CA] * get_tpm("PIK3CA", cell_line, source) +
            x[C.w_PIK3CB] * get_tpm("PIK3CB", cell_line, source) +
            x[C.w_PIK3CD] * get_tpm("PIK3CD", cell_line, source) +
            x[C.w_PIK3CG] * get_tpm("PIK3CG", cell_line, source))


def init_PTEN(x, cell_line, source):
    return x[C.w_PTEN] * get_tpm("PTEN", cell_line, source)


def init_R(x, cell_line, source):
    return (x[C.w_RASA1] * get_tpm("RASA1", cell_line, source) +
            x[C.w_RASA2] * get_tpm("RASA2", cell_line, source) +
            x[C.w_RASA3] * get_tpm("RASA3", cell_line, source))


def init_O(x, cell_line, source):
    return (x[C.w_SOS1] * get_tpm("SOS1", cell_line, source) +
            x[C.w_SOS2] * get_tpm("SOS2", cell_line, source))


def init_A(x, cell_line, source):
    return x[C.w_A] * get_tpm("GAB1", cell_line, source)


def init_Akt(x, cell_line, source):
    return (x[C.w_AKT1] * get_tpm("AKT1", cell_line, source) +
            x[C.w_AKT2] * get_tpm("AKT2", cell_line, source))


def init_RsD(x, cell_line, source):
    return (x[C.w_HRAS] * get_tpm("HRAS", cell_line, source) +
            x[C.w_KRAS] * get_tpm("KRAS", cell_line, source) +
            x[C.w_NRAS] * get_tpm("NRAS", cell_line, source))


def init_Raf(x, cell_line, source):
    return (x[C.w_ARAF] * get_tpm("ARAF", cell_line, source) +
            x[C.w_BRAF] * get_tpm("BRAF", cell_line, source) +
            x[C.w_RAF1] * get_tpm("RAF1", cell_line, source))



def init_MEK(x, cell_line, source):
    return (x[C.w_MAP2K1] * get_tpm("MAP2K1", cell_line, source) +
            x[C.w_MAP2K2] * get_tpm("MAP2K2", cell_line, source))


def init_T(x, cell_line, source):
    return x[C.w_T] * get_tpm("PTPN1", cell_line, source)


def init_CREBn(x, cell_line, source):
    return x[C.w_CREBn] * get_tpm("CREB1", cell_line, source)


def init_ERKc(x, cell_line, source):
    return (x[C.w_MAPK1] * get_tpm("MAPK1", cell_line, source) +
            x[C.w_MAPK3] * get_tpm("MAPK3", cell_line, source))


def init_Elk1n(x, cell_line, source):
    return x[C.w_Elk1n] * get_tpm("ELK1", cell_line, source)


def init_RSKc(x, cell_line, source):
    return (x[C.w_RPS6KA1] * get_tpm("RPS6KA1", cell_line, source) +
            x[C.w_RPS6KA2] * get_tpm("RPS6KA2", cell_line, source) +
            x[C.w_RPS6KA3] * get_tpm("RPS6KA3", cell_line, source) +
            x[C.w_RPS6KA6] * get_tpm("RPS6KA6", cell_line, source))


def tpm2ival(x, y0, cell_line, source='CCLE'):
    x[C.PTEN] *= init_PTEN(x, cell_line, source)

    y0[V.E1] *= init_E1(x, cell_line, source)
    y0[V.E2] *= init_E2(x, cell_line, source)
    y0[V.E3] *= init_E3(x, cell_line, source)
    y0[V.E4] *= init_E4(x, cell_line, source)
    y0[V.G] *= init_G(x, cell_line, source)
    y0[V.S] *= init_S(x, cell_line, source)
    y0[V.I] *= init_I(x, cell_line, source)
    y0[V.R] *= init_R(x, cell_line, source)
    y0[V.O] *= init_O(x, cell_line, source)
    y0[V.A] *= init_A(x, cell_line, source)
    y0[V.Akt] *= init_Akt(x, cell_line, source)
    y0[V.RsD] *= init_RsD(x, cell_line, source)
    y0[V.Raf] *= init_Raf(x, cell_line, source)
    y0[V.MEK] *= init_MEK(x, cell_line, source)
    y0[V.T] *= init_T(x, cell_line, source)
    y0[V.CREBn] *= init_CREBn(x, cell_line, source)
    y0[V.ERKc] *= init_ERKc(x, cell_line, source)
    y0[V.Elk1n] *= init_Elk1n(x, cell_line, source)
    y0[V.RSKc] *= init_RSKc(x, cell_line, source)

    return x, y0


def mul_ratio2mcf7(x, y0, patientID, source='CCLE'):
    x[C.PTEN] *= init_PTEN(x, patientID, source) / init_PTEN(x, 'MCF7', source)

    y0[V.E1] *= init_E1(x, patientID, source) / init_E1(x, 'MCF7', source)
    y0[V.E2] *= init_E2(x, patientID, source) / init_E2(x, 'MCF7', source)
    y0[V.E3] *= init_E3(x, patientID, source) / init_E3(x, 'MCF7', source)
    y0[V.E4] *= init_E4(x, patientID, source) / init_E4(x, 'MCF7', source)
    y0[V.G] *= init_G(x, patientID, source) / init_G(x, 'MCF7', source)
    y0[V.S] *= init_S(x, patientID, source) / init_S(x, 'MCF7', source)
    y0[V.I] *= init_I(x, patientID, source) / init_I(x, 'MCF7', source)
    y0[V.R] *= init_R(x, patientID, source) / init_R(x, 'MCF7', source)
    y0[V.O] *= init_O(x, patientID, source) / init_O(x, 'MCF7', source)
    y0[V.A] *= init_A(x, patientID, source) / init_A(x, 'MCF7', source)
    y0[V.Akt] *= init_Akt(x, patientID, source) / init_Akt(x, 'MCF7', source)
    y0[V.RsD] *= init_RsD(x, patientID, source) / init_RsD(x, 'MCF7', source)
    y0[V.Raf] *= init_Raf(x, patientID, source) / init_Raf(x, 'MCF7', source)
    y0[V.MEK] *= init_MEK(x, patientID, source) / init_MEK(x, 'MCF7', source)
    y0[V.T] *= init_T(x, patientID, source) / init_T(x, 'MCF7', source)
    y0[V.CREBn] *= init_CREBn(x, patientID, source) / init_CREBn(x, 'MCF7', source)
    y0[V.ERKc] *= init_ERKc(x, patientID, source) / init_ERKc(x, 'MCF7', source)
    y0[V.Elk1n] *= init_Elk1n(x, patientID, source) / init_Elk1n(x, 'MCF7', source)
    y0[V.RSKc] *= init_RSKc(x, patientID, source) / init_RSKc(x, 'MCF7', source)

    return x, y0